### ISSUE TITLE
Create GUI for sshnoports for windows/linux/macos

### ISSUE THREAD

**Issue by cconstab** [atsign-foundation/noports#163]
```[tasklist]
### MVP Release tasks
- [x] 1. GUI config file management
- [x] 1.1. syncing using atServers (with a last updated time to prevent conflicts)
- [x] 1.2. Import/exporting from local filesystem to sandbox
- [x] 2. Talk to Gary
- [x] 2.1. Pure Dart SSH (@gkc as part of v4.0.0 changes)
- [x] 2.2. Implement Background Sessions in GUI
- [x] 2.2.1. Lengthen timeout for sshnp gui ~
- [x] 3. Error Messages
- [x] 3.1. When bad connection info is supplied
- [x] 3.2. When daemon timeout / is offline
- [ ] 4. Unit tests for sshnp
- [x] 5. GUI private key file management
```

A partner requires a GUI for sshnp, that can also forward ports without any ports being needed to be opened on the device or firewalls.

- [x]  Create MVP
- [x] Design Nicer GUI
- [x] Code final product
- [ ] Automate testing
- [ ] Launch

Some useful material in [this document](https://docs.google.com/presentation/d/1HVGq-SLVst3S06WLvXC1cswBPSOAyTGfkyOpuy3uTnI/edit#slide=id.g24f28d34160_0_175)

Key enabler: since advent of the rendezvous service, sshd does not need to run on the client. Therefore we can run a client on any platform as we just need an ssh client

Functionality of the GUI needs to include something like the "Tunnel" section of the Putty app (Source, Destination, and local or remote)

**Comment by gkc** [comment]
Setting priority to P0 as this needs to get done during this sprint. Removing the SP estimate so we remember to discuss during sprint planning

**Comment by Dariamarg** [comment]
[Figma](https://www.figma.com/file/evF7SU90OFkobkwtcUIZsX/GUI?type=design&node-id=0%3A1&mode=design&t=sUs19EX6WKzgzpWN-1) Here's where I'm working!

**Comment by xavierchanth** [comment]
A little more in the ways of the base code for sshnp to get this over the hump, but config files are there which means we should be able to get started.

Ideally the GUI reads `~/.sshnp/config/*` for all of the various config files saved to the device, then only gets the ones where "FROM" matches the currently onboarded atSign. Alternatively (or in addition), these same config files are stored on the atServer so that your config follows you to your other machines.

Maybe the GUI saves both locally and to the atServer, that way you can always switch to terminal if that's what you prefer. Or there is a toggle for "sync with other devices" which can be added as a parameter in the config file.

**Comment by gkc** [comment]
@CurtlyCritchlow Please can you provide an update on this ticket and also [this other ticket ](https://github.com/atsign-foundation/at_protocol/issues/92)

**Comment by CurtlyCritchlow** [comment]
Did some UI work but waiting on some additional update to the sshnp package to connect the UI to the backend. @XavierChanth will be working on the update this sprint.

<img width="715" alt="image" src="https://github.com/atsign-foundation/at_protocol/assets/44856451/289d7ee3-7d88-47d5-97d4-50bf9234f41e">

**Comment by xavierchanth** [comment]
PR #306 provides the necessary methods to read in all of the config files which contain this information. I will also write the code to write a file for the add new connection

**Comment by CurtlyCritchlow** [comment]
Update: we can ssh from the GUI app. To complete the MVP we have to display the string printed to stdout upon success in the UI.

**Comment by xavierchanth** [comment]
Increasing the SPs as I will be focusing on this in PR70

**Comment by xavierchanth** [comment]
A short task list for my focuses:

- [x] Refactor home screen controllers to enable proper file management
- [x] Implement terminal screen controllers to enable multiple terminal windows (terminal process/ssh instance management)
- [x] Implement dynamic navigation controller to enable multiple terminal navigation (UI navigation management)
- [ ] Pure Dart SSH

**Comment by CurtlyCritchlow** [comment]
[UI design](https://www.figma.com/file/evF7SU90OFkobkwtcUIZsX/Desktop-App?node-id=784%3A7355&mode=dev) of the key management has been completed. The design implementation was started in PR74 and will continue in PR75.

**Comment by CurtlyCritchlow** [comment]
The UI design for macOS app has been mostly completed. Pending is the onboarding screen. The next activities are to complete the onboarding/landing screen, unit test, widget test and update the app to use the latest version of sshnoports.

**Comment by CurtlyCritchlow** [comment]
The above mentioned activities are still in progress.

**Comment by CurtlyCritchlow** [comment]
The onboarding screen has been completed, Private key management functionality has been added. Pending activity is to pass the private key properties to the dart.ssh client to ssh to a remote device.

**Comment by ksanty** [comment]
Hi @CurtlyCritchlow we have a launch date of Jan 19th on the GUI, is that still accurate?  :-)

**Comment by CurtlyCritchlow** [comment]
Looking to pushed to test flight by Jan 17th. We'll still have to do user testing, so not ready for an official launch to production but would be ready for early adopters if the team decide to make it available to them.

**Comment by CurtlyCritchlow** [comment]
App was pushed to test flight for testing. This sprint it will be pushed to windows dashboard for testing (already started). Updates and big fixes based on preliminary testing are. Adding keybindings, make hung shell non blocking

**Comment by ksanty** [comment]
Awesome @CurtlyCritchlow congrats!

**Comment by Dariamarg** [comment]
UI/UX fixes sent out durning last sprint, working through post onboarding flow ('how to get started')

**Comment by xavierchanth** [comment]
Duplicate of #717
